File: backend/src/main/java/com/meuprojeto/user/UserService.java
package com.seuprojeto.user;

import com.seuprojeto.user.dto.*;
import lombok.RequiredArgsConstructor;
import org.springframework.security.core.userdetails.UserDetails;
import org.springframework.security.core.userdetails.UserDetailsService;
import org.springframework.security.core.userdetails.UsernameNotFoundException;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;

@Service
@RequiredArgsConstructor
public class UserService implements UserDetailsService {

    private final UserRepository userRepository;
    private final PasswordEncoder passwordEncoder;

    @Override
    public UserDetails loadUserByUsername(String email) throws UsernameNotFoundException {
        return userRepository.findByEmail(email)
                .orElseThrow(() -> new UsernameNotFoundException("Usuário não encontrado: " + email));
    }

    public User register(RegisterRequest request) {
        if (userRepository.existsByEmail(request.getEmail())) {
            throw new IllegalArgumentException("Já existe um usuário cadastrado com este e-mail");
        }
        User user = User.builder()
                .nome(request.getNome())
                .email(request.getEmail())
                .senha(passwordEncoder.encode(request.getSenha()))
                .role(Role.USER)
                .build();
        return userRepository.save(user);
    }

    public User findByEmail(String email) {
        return userRepository.findByEmail(email)
                .orElseThrow(() -> new UsernameNotFoundException("Usuário não encontrado: " + email));
    }

    public User updateProfile(String email, UpdateProfileRequest request) {
        User user = findByEmail(email);
        user.setNome(request.getNome());
        return userRepository.save(user);
    }

    public UserProfileResponse toProfileResponse(User user) {
        return UserProfileResponse.builder()
                .id(user.getId())
                .nome(user.getNome())
                .email(user.getEmail())
                .role(user.getRole().name())
                .build();
    }
}

File: backend/src/user/user.service.ts
import { ConflictException, Injectable, NotFoundException } from '@nestjs/common';
import * as bcrypt from 'bcryptjs';
import { PrismaService } from '../prisma/prisma.service';
import { RegisterDto } from '../auth/dto/register.dto';
import { UpdateProfileDto } from './dto/update-profile.dto';
import { Role, User } from '@prisma/client';

@Injectable()
export class UserService {
  constructor(private readonly prisma: PrismaService) {}

  async findByEmailOrNull(email: string): Promise<User | null> {
    return this.prisma.user.findUnique({ where: { email } });
  }

  async findByEmail(email: string): Promise<User> {
    const user = await this.findByEmailOrNull(email);
    if (!user) {
      throw new NotFoundException(`Usuário não encontrado: ${email}`);
    }
    return user;
  }

  async register(dto: RegisterDto): Promise<User> {
    const jaExiste = await this.prisma.user.findUnique({ where: { email: dto.email } });
    if (jaExiste) {
      throw new ConflictException('Já existe um usuário cadastrado com este e-mail');
    }
    const senhaHash = await bcrypt.hash(dto.senha, 10);
    return this.prisma.user.create({
      data: {
        nome: dto.nome,
        email: dto.email,
        senha: senhaHash,
        role: Role.USER,
      },
    });
  }

  async updateProfile(email: string, dto: UpdateProfileDto): Promise<User> {
    const user = await this.findByEmail(email);
    return this.prisma.user.update({
      where: { id: user.id },
      data: { nome: dto.nome },
    });
  }

  toProfileResponse(user: User) {
    return {
      id: user.id,
      nome: user.nome,
      email: user.email,
      role: user.role,
    };
  }
}
